password-cli/app/Main.hs

```diff
@@ -0,0 +1,209 @@
+import Data.Password.Types
+import Data.Text (Text)
+import qualified Data.Text.IO as T
+import Options.Applicative
+import Paths_password_cli (version)
+import System.Environment (getArgs)
+import System.Exit (exitFailure)
+import System.IO (IOMode(ReadMode), hSetEcho, stdin, stderr, withFile)
+
+main :: IO ()
+main =
+  execParserPure defaultPrefs cliOpts . defaultHelp <$> getArgs
+  >>= handleParseResult
+  >>= runCmd
+  where defaultHelp =
+          \case
+            [] -> ["help"]
+            xs -> xs
+
+data Cmd
+  = HashCmd HashOpts
+  | CheckCmd CheckOpts
+  | HelpCmd (Maybe String)
+
+data HashOpts = HashOpts {
+    quiet :: Bool,
+    password :: Maybe FilePath,
+    hashAlgorithm :: HashAlgorithm
+  }
+
+data CheckOpts = CheckOpts {
+    hash :: Either FilePath Text,
+    quiet :: Bool,
+    password :: Maybe FilePath,
+    checkAlgorithm :: CheckAlgorithm
+  }
+
+data HashAlgorithm
+  = PBKDF2HashAlgo PBKDF2.PBKDF2Params
+  | BcryptHashAlgo Int
+  | ScryptHashAlgo Scrypt.ScryptParams
+  | Argon2HashAlgo Argon2.Argon2Params
+
+data CheckAlgorithm
+  = PBKDF2CheckAlgo
+  | BcryptCheckAlgo
+  | ScryptCheckAlgo
+  | Argon2CheckAlgo
+
+cliOpts :: ParserInfo Cmd
+cliOpts = info commandsParser (fullDesc <> header ("Password CLI " <> showVersion version))
+  where
+    commandsParser :: Parser Cmd
+    commandsParser =
+      hsubparser
+        ( command
+            "hash"
+            (info (HashCmd <$> hashOptsParser) (progDesc "hash password"))
+            <> command
+              "check"
+              (info (CheckCmd <$> checkOptsParser) (progDesc "check hashed password"))
+            <> command "help" (info (HelpCmd <$> optional (argument str (metavar "COMMAND")) <**> helper) (progDesc "Show command help"))
+        )
+        <**> helper
+    hashOptsParser :: Parser HashOpts
+    hashOptsParser =
+      HashOpts
+        <$> switch (short 'q' <> long "quiet")
+        <*> optional (option str (metavar "PASSWORD-FILE" <> long "password-file"))
+        <*> algorithmHashParser
+    checkOptsParser :: Parser CheckOpts
+    checkOptsParser =
+      CheckOpts
+        <$> (   (Right <$> option str (metavar "HASH" <> long "hash"))
+            <|> (Left <$> option str (metavar "HASH-FILE" <> long "hash-file")))
+        <*> switch (short 'q' <> long "quiet")
+        <*> optional (option str (metavar "PASSWORD-FILE" <> long "password-file"))
+        <*> algorithmCheckParser
+    algorithmHashParser :: Parser HashAlgorithm
+    algorithmHashParser =
+      hsubparser
+        ( command "argon2" (info (Argon2HashAlgo <$> algoParser argon2Def) (progDesc "Argon2"))
+            <> command "bcrypt" (info (BcryptHashAlgo <$> algoParser bcryptDef) (progDesc "Bcrypt"))
+            <> command "pbkdf2" (info (PBKDF2HashAlgo <$> algoParser pbkdf2Def) (progDesc "PBKDF2"))
+            <> command "scrypt" (info (ScryptHashAlgo <$> algoParser scryptDef) (progDesc "Scrypt"))
+        )
+        <**> helper
```

Comment:
The `hsubparser` already adds the `<**> helper`, so this will result in the following output, when doing `password-cli hash --help`

```
Available options:
  -h,--help                Show this help text
  -h,--help                Show this help text

Available commands:
  argon2                   Argon2
  bcrypt                   Bcrypt
  pbkdf2                   PBKDF2
  scrypt                   Scrypt
```

Reply:
same in the `algorithmCheckParser`